src/tango/ip/fd_netlink.c

```diff
@@ -0,0 +1,760 @@
+    close( fd );
+    return -1;
+  }
+
+  return fd;
+}
+
+
+void
+fd_nl_close_socket( int fd ) {
+  if( fd >= 0 ) {
+    close( fd );
+  }
+}
+
+int
+fd_nl_init( fd_nl_t * nl, uint seq ) {
+  nl->seq = seq;
+  nl->fd  = fd_nl_create_socket();
+
+  /* returns 1 for failure, 0 for success */
+  return nl->fd < 0 ? 1 : 0;
+}
+
+void
+fd_nl_fini( fd_nl_t * nl ) {
+  fd_nl_close_socket( nl->fd );
+}
+
+
+long
+fd_nl_load_route_table( fd_nl_t *             nl,
+                        fd_nl_route_entry_t * route_table,
+                        ulong                 route_table_cap ) {
+  int fd = nl->fd;
+  if( fd < 0 ) return -1;
+
+  /* format the request */
+
+  /* Request struct */
+  struct {
+    struct nlmsghdr nlh;  /* Netlink header */
+    struct rtmsg    rtm;  /* Payload - route message */
+  } nl_request;
+
+  fd_memset( &nl_request, 0, sizeof( nl_request ) );
+
+  uint seq = nl->seq++;
+
+  nl_request.nlh.nlmsg_type  = RTM_GETROUTE;  /* We wish to get routes */
+#if 0
+  /* CAP_NET_ADMIN required for NLM_F_ATOMIC */
+  nl_request.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_ROOT | NLM_F_ATOMIC;
+#else
+  nl_request.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_ROOT;
+#endif
+  nl_request.nlh.nlmsg_len   = sizeof(nl_request);
+  nl_request.nlh.nlmsg_pid   = 0;
+  nl_request.nlh.nlmsg_seq   = seq;
+
+  /* request only IPv4 routes */
+  nl_request.rtm.rtm_family  = AF_INET;
+
+  long sent = send( fd, &nl_request, sizeof(nl_request), 0 );
+  if( sent == -1 ) {
+    FD_LOG_WARNING(( "Unable to make netlink request. Error: %d %s", errno, strerror( errno ) ));
+    return -1;
+  }
+
+  /* receiving */
+  long len;
+
+  /* set alignment such that the returned data is aligned */
+  uchar __attribute__(( aligned(16) )) ibuf[4096]   = {0};
+
+  /* Pointer to the messages head */
+  struct nlmsghdr *h = (struct nlmsghdr *)ibuf;
+
+  /* Try to read the message in case of busy or interrupted call */
+  while(1) {
+    do {
+      len = recv( fd, ibuf, sizeof(ibuf), 0 );
+    } while( len < 0 && (errno == EINTR || errno == EAGAIN) );
+
+    if( len < 0 ) {
+      FD_LOG_WARNING(( "netlink recv failed with %d %s", errno, strerror( errno ) ));
+      return -1;
+    }
+
+    if( h->nlmsg_seq == seq ) break;
+  }
+
+  /* index into route_entries */
+  long route_entry_idx = 0L;
+
+  /* interpret the response */
+
+  long msglen = len;
+
+  /* Iterate through all messages in buffer */
+
+  while( NLMSG_OK(h, msglen) ) {
+    /* are we still within the bounds of the table? */
+    if( route_entry_idx >= (long)route_table_cap ) {
+      /* we filled the table */
+      return -1; /* return failure */
+    }
+
+    /* current route entry */
+    fd_nl_route_entry_t * entry = route_table + route_entry_idx;
+
+    /* clear current entry */
+    fd_memset( entry, 0, sizeof( *entry ) );
+
+    if( h->nlmsg_flags & NLM_F_DUMP_INTR ) {
+      /* function was interrupted - don't switch to new routing table */
+
+      return -1; /* return failure */
```

Comment:
Don't love this error code here. Most of these errors are unrecoverable and we would want to FD_LOG_ERR and crash in response, but this is a benign retryable condition. Can you separate the codes with a good note about this in the header comments?

Reply:
I originally had this returning void. I decided to return the number of entries, and forgot to add named constants. Will fix

Reply:
I kept -1 as the only error. It's assumed to be transient, and that's in the comments in the header. Any known non-transients are handled with FD_LOG_ERR instead